File: rw-querydsl/src/main/java/com/example/rwquerydsl/common/config/SecurityConfig.java
package com.example.rwquerydsl.common.config;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter;
import com.example.rwquerydsl.security.CustomAccessDeniedHandler;
import com.example.rwquerydsl.security.CustomAuthenticationEntryPoint;
import com.example.rwquerydsl.security.JwtCustomFilter;
import com.example.rwquerydsl.security.JwtCustomProvider;

@EnableWebSecurity
public class SecurityConfig {

    private final AuthenticationManagerBuilder authenticationManagerBuilder;
    private final ConfigProvider configProvider;

    @Autowired
    public SecurityConfig(
        AuthenticationManagerBuilder authenticationManagerBuilder,
        JwtCustomProvider jwtCustomProvider,
        ConfigProvider configProvider
    ) {
        this.authenticationManagerBuilder = authenticationManagerBuilder;
        this.configProvider = configProvider;
        authenticationManagerBuilder.authenticationProvider(jwtCustomProvider);
    }

    @Bean
    public BCryptPasswordEncoder bCryptPasswordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
        http.setSharedObject(AuthenticationManager.class, this.authenticationManagerBuilder.getOrBuild());
        http.csrf()
            .disable()
            .authorizeRequests()
            .antMatchers("/user/**")
            .hasAuthority(configProvider.getJwtRole())
            .antMatchers("/profiles/{username}/**")
            .hasAuthority(configProvider.getJwtRole())
//            .antMatchers("/check/**")
//            .hasAuthority(configProvider.getBasicRole())
            .anyRequest()
            .permitAll()
            .and()
            .addFilterAfter(new JwtCustomFilter(http.getSharedObject(AuthenticationManager.class)), WebAsyncManagerIntegrationFilter.class)
            //            .addFilterAfter(new JwtCustomFilter(this.authenticationManagerBuilder.getOrBuild()), WebAsyncManagerIntegrationFilter.class)
            .httpBasic().disable()
            .requestCache().disable()
            .securityContext().disable()
            .sessionManagement().disable()
            .formLogin().disable()
            .logout().disable()
            .headers().disable()
            .exceptionHandling().authenticationEntryPoint(new CustomAuthenticationEntryPoint())
                                .accessDeniedHandler(new CustomAccessDeniedHandler());
        //  .exceptionHandling().disable();
        //  .anonymous().disable();
        return http.build();
    }
}

File: spring-boot-kt/src/main/kotlin/real/world/springbootkt/global/config/security/SecurityConfig.kt
package real.world.springbootkt.global.config.security

import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter
import real.world.springbootkt.global.common.SecurityProperties
import real.world.springbootkt.global.filter.JwtFilter

@Configuration
@EnableWebSecurity
class SecurityConfig(
    private val securityProperties: SecurityProperties,
    private val objectMapper: ObjectMapper,
    private val authenticationManagerBuilder: AuthenticationManagerBuilder,
    jwtProvider: JwtProvider
) {
    init {
        authenticationManagerBuilder.authenticationProvider(jwtProvider)
    }

    @Bean
    fun filterChain(http: HttpSecurity): SecurityFilterChain {
        http.setSharedObject(
            AuthenticationManager::class.java,
            this.authenticationManagerBuilder.getOrBuild()
        )
        return http.csrf { it.disable() }
            .cors { it.disable() }
            .httpBasic { it.disable() }
            .requestCache { it.disable() }
            .securityContext { it.disable() }
            .sessionManagement { it.disable() }
            .formLogin { it.disable() }
            .logout { it.disable() }
            .headers { it.disable() }
            .addFilterAfter(
                JwtFilter(http.getSharedObject(AuthenticationManager::class.java)),
                WebAsyncManagerIntegrationFilter::class.java
            )
            .authorizeHttpRequests {
                it.requestMatchers("/user/**")
                    .hasAuthority(securityProperties.jwtRole)
                    .requestMatchers("/profiles/{username}/**")
                    .hasAuthority(securityProperties.jwtRole)
                    .anyRequest()
                    .permitAll()
            }
            .exceptionHandling {
                it.authenticationEntryPoint(CustomAuthenticationEntryPoint(objectMapper))
                    .accessDeniedHandler(CustomAccessDeniedHandler(objectMapper))
            }
            .build()
    }
}